drivers/entropy/entropy_w91_trng.c

```diff
@@ -0,0 +1,128 @@
+/*
+ * Copyright (c) 2024 Telink Semiconductor
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT telink_w91_trng
+
+#include <zephyr/drivers/entropy.h>
+#include <ipc/ipc_based_driver.h>
+
+enum {
+	IPC_DISPATCHER_TRNG_GET_ENTROPY = IPC_DISPATCHER_ENTROPY_TRNG,
+};
+
+struct entropy_w91_config {
+	uint8_t instance_id;            /* instance id */
+};
+
+struct entropy_w91_data {
+	struct ipc_based_driver ipc;    /* ipc driver part */
+};
+
+struct entropy_w91_trng_get_entropy_resp {
+	int err;
+	uint16_t length;
+	uint8_t *buffer;
+};
+
+/* API implementation: driver initialization */
+static int entropy_w91_trng_init(const struct device *dev)
+{
+	struct entropy_w91_data *data = dev->data;
+
+	ipc_based_driver_init(&data->ipc);
+
+	return 0;
+}
+
+/* APIs implementation: get entropy */
+static size_t pack_entropy_w91_trng_get_entropy(
+	uint8_t inst, void *unpack_data, uint8_t *pack_data)
+{
+	uint16_t *p_length = unpack_data;
+	size_t pack_data_len = sizeof(uint32_t) + sizeof(*p_length);
+
+	if (pack_data != NULL) {
+		uint32_t id = IPC_DISPATCHER_MK_ID(IPC_DISPATCHER_TRNG_GET_ENTROPY, inst);
+
+		IPC_DISPATCHER_PACK_FIELD(pack_data, id);
+		IPC_DISPATCHER_PACK_FIELD(pack_data, *p_length);
+	}
+
+	return pack_data_len;
+}
+
+static void unpack_entropy_w91_trng_get_entropy(void *unpack_data,
+		const uint8_t *pack_data, size_t pack_data_len)
+{
+	struct entropy_w91_trng_get_entropy_resp *p_trng_get_entropy_resp = unpack_data;
+
+	pack_data += sizeof(uint32_t);
+	IPC_DISPATCHER_UNPACK_FIELD(pack_data, p_trng_get_entropy_resp->err);
+	IPC_DISPATCHER_UNPACK_FIELD(pack_data, p_trng_get_entropy_resp->length);
+
+	size_t expect_len = sizeof(uint32_t) + sizeof(p_trng_get_entropy_resp->err) +
+			sizeof(p_trng_get_entropy_resp->length) + p_trng_get_entropy_resp->length;
+
+	if (expect_len != pack_data_len) {
+		p_trng_get_entropy_resp->err = -EINVAL;
+		return;
+	}
+
+	IPC_DISPATCHER_UNPACK_ARRAY(pack_data, p_trng_get_entropy_resp->buffer,
+			p_trng_get_entropy_resp->length);
+}
+
+static int entropy_w91_trng_get_entropy(const struct device *dev,
+					uint8_t *buffer, uint16_t length)
+{
+	struct entropy_w91_trng_get_entropy_resp trng_get_entropy_resp = {
+		.buffer	= buffer,
+	};
+
+	struct ipc_based_driver *ipc_data = &((struct entropy_w91_data *)dev->data)->ipc;
+	uint8_t inst = ((struct entropy_w91_config *)dev->config)->instance_id;
+
+	IPC_DISPATCHER_HOST_SEND_DATA(ipc_data, inst,
+		entropy_w91_trng_get_entropy, &length, &trng_get_entropy_resp,
+		CONFIG_ENTROPY_TELINK_W91_IPC_RESPONSE_TIMEOUT_MS);
+
+	return trng_get_entropy_resp.err;
+}
+
+/* API implementation: get_entropy_isr */
+static int entropy_w91_trng_get_entropy_isr(const struct device *dev,
+					    uint8_t *buffer, uint16_t length,
+					    uint32_t flags)
+{
+	ARG_UNUSED(flags);
+
+	/* No specific handling in case of running from ISR, just call standard API */
+	return entropy_w91_trng_get_entropy(dev, buffer, length);
```

Comment:
It won't work: we are in ISR here.
There is no possibility to receive IPC response in ISR.
To overcome this I propouse to use standard c random here.
Init standard c random when we call it first time using 10MHz MTIMER value.
But at first plese measure timings of these function execition using ligic analyzer to ensure that we can spend this time in ISR.

Reply:
This approach will create delays at large values (if the user requests 4K bytes), so it is better to specify that getting entropy is not supported in interrupts.